File: Exercises2024/Chapter06_vle_free_energy/Fortran/1-Gibbs/Source/sample.f
      Subroutine Sample(I, En, Vir)

C     Writes Quantities To File

      Implicit None

      Include 'parameter.inc'
      Include 'conf.inc'
      Include 'system.inc'

      Integer I, BoxID
      Double Precision En(*), Enp(2), Vir(*), Press(2), Vol, 
     &                 Rho(2)

      Do BoxID = 1, 2
         Vol = Box(BoxID)**3
         Rho(BoxID) = Dble(Npbox(BoxID))/Vol
         Press(BoxID) = Rho(BoxID)/Beta + Vir(BoxID)/(3.D0*Vol)
         
         If (Npbox(BoxID).Ne.0) Then
            Enp(BoxID) = En(BoxID)/Dble(Npbox(BoxID))
         Else
            Enp(BoxID) = 0.D0
         End If
      End Do
      Write (66,'(7e20.10)') Dble(I), Dble(Enp(1)), Dble(Enp(2)), Dble(Press(1)), 
     &              Dble(Press(2)), Dble(Rho(1)), Dble(Rho(2))
!      Write (44,'(2(I6,F10.2))') Npbox(1), Box(1)**3, Npbox(2), Box(2)
!     &                            **3
      Write(45,'(3e20.10)') Dble(I),(Dble(Npbox(1))/Box(1)**3),
     &     (Dble(Npbox(2))/Box(2)**3)
      Return
      End

File: Exercises2024/Chapter06_vle_free_energy/C/1-Gibbs/Source/sample.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>
#include "conf.h"
#include "system.h"

void Sample(int I,double *En,double *Vir)
{
  //C     Writes Quantities To File
  int BoxID;
  double Enp[2], Press[2], Vol, Rho[2];
  FILE* fileptr;
  
  for(BoxID = 0; BoxID<2;BoxID++)
    {
      Vol = pow(Box[BoxID],3);
      Rho[BoxID] = (double)(Npbox[BoxID])/Vol;
      Press[BoxID] = Rho[BoxID]/Beta + Vir[BoxID]/(3.0*Vol);
         
      if (Npbox[BoxID]!=0)
	{
	  Enp[BoxID] = En[BoxID]/(double)(Npbox[BoxID]);
	}
      else
	{
	  Enp[BoxID] = 0.;
	}
    }
  
  fileptr=fopen("fort.66","a");
  fprintf(fileptr,"%d\t%lf\t%lf\t%lf\t%lf\t%lf\t%lf\n",I, Enp[0],Enp[1],Press[0],Press[1],Rho[0],Rho[1]);
  fclose(fileptr);
  fileptr=fopen("fort.44","a");
  fprintf(fileptr,"%d\t%lf\t%d\t%lf\n",Npbox[0],pow(Box[0],3),Npbox[1],pow(Box[1],3));
  fclose(fileptr);
  fileptr=fopen("fort.45","a");
  fprintf(fileptr,"%d\t%lf\t%lf\n",I,Npbox[0]/pow(Box[0],3),Npbox[1]/pow(Box[1],3));
  fclose(fileptr);
  return;
}
